File: problems_java/leetCode18.java
package problems_java;

import java.util.ArrayList;
import java.util.Arrays;

public class leetCode18 {
    public List<List<Integer>> fourSum(int[] nums, int target){
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        for ( int i =  0 ; i < nums.length; i++){
            if (i > 0 && nums[i] == nums[i - 1]){
                continue;
            }
            for (int j = i + 1; j < nums.length; j ++){
                if(j > i+1  && nums[j] == nums[j-1]){
                    continue;
                }
                int left = j + 1;
                int right = nums.length - 1;
                while (right > left){
                    int sum = nums[i] + nums[j] + nums[left] + nums[right];
                    if (sum > target ){
                        right -= 1;
                    }
                    else if (sum < 0 ){
                        left += 1;
                    }
                    else{
                        result.append(Arrays.asList([nums[i],nums[j],nums[left],nums[right]));
                        while(left < right && nums[left] == nums[left + 1]){
                            left += 1;
                        }
                        while(left < right && nums[right] == nums[right-1]){
                            right -= 1;
                        }
                        left += 1;
                        right -= 1;

                    }
                }
            }
        }
        return result;
    }
    public static void main(String[] args){
    }
}

File: mid/leetCode18.py
#18. 4Sum
def fourSum(nums,target):
    nums.sort()
    n = len(nums)
    result = []

    for i in range(n):
        if i > 0 and nums[i] == nums[i-1]: continue
        for k in range(i + 1,n):
            if k > i + 1 and nums[k] == nums[k -1]:continue
            left = k + 1 
            right = n - 1 
            while left < right:
                if nums[i] + nums[k] + nums[left] + nums[right] > target : right -= 1
                elif nums[i] + nums[k] + nums[left] + nums[right] < target : left += 1
                else:
                    result.append([nums[i],nums[k],nums[left],nums[right]])
                    while left < right and nums[left] == nums[left + 1]: left += 1
                    while left < right and nums[right] == nums[right - 1]: right -= 1
                    left += 1
                    right -= 1
    print(result)
    return result 


nums = [1,0,-1,0,-2,2]
target = 0
fourSum(nums,target)